Documentation/proposals/202305-scrapeclasses.md

```diff
@@ -0,0 +1,217 @@
+- Allow for a user to select a configuration snippet by name in their probe/podmonitor/servicemonitor endpoint configuration.
+- Avoid giving the user the ability to use arbirary files within the Prometheus pod.
+
+### Audience
+
+- Users who serve Prometheus as a service and want to give their customers autonomy in defining monitors, but want to provide a default configuration for scraping.
+
+## How
+
+The proposed solution is to introduce a notion of a *scrape class*, akin to a Kubernetes [storage class](https://kubernetes.io/docs/concepts/storage/storage-classes/).
+A scrape class defines TLS settings (and possibly other settings in future, e.g. sensitive authorization settings) to be applied to all scrape configs of that class.
+
+One scrape class may be designated as the default class, in which case that class is applied to any scrape config that doesn't specify a value for `scrapeClass`.
+
+When defining a podmonitor/servicemonitor/probe/scrapeconfig, a user may assign a scrape class via the `scrapeClass` field.
+For monitors, each endpoint may be assigned a scrape class.
+
+Class names are assumed to be installation-specific. In practice, some common class names like `istio-mtls` are likely to emerge.
+
+### Prometheus Resource
+
+It is proposed that the `Prometheus` and `PrometheusAgent` resources contain a new section for defining scrape classes.
+
+The rationale for defining scrape classes inline is that, in practice, the TLS file paths are closely related to the `volumeMounts`
+of the `Prometheus` spec. An alternative is outlined later, of factoring the class definitions into a separate resource.
+
+```yaml
+apiVersion: monitoring.coreos.com/v1
+kind: Prometheus
+spec:
+  # define scrape classes for use by the monitors
+  # one class may be designated as the default class
+  scrapeClasses:
+    - name: istio-mtls
+      default: true
+      tlsConfig:
+        caFile: "/etc/istio-certs/root-cert.pem"
+        certFile: "/etc/istio-certs/cert-chain.pem"
+        keyFile: "/etc/istio-certs/key.pem"
+        insecureSkipVerify: true
+
+  # mount the certs from the istio sidecar (shown here for illustration purposes)
+  volumeMounts:
+    - name: istio-certs
+      mountPath: "/etc/istio-certs/"
+```
+
+Any object references in the scrape class definition are assumed to refer to objects in the namespace of the `Prometheus` object.
+
+### PodMonitor Resource
+
+Allow the user to select a scrape class for each endpoint.
+
+```yaml
+apiVersion: monitoring.coreos.com/v1
+kind: PodMonitor
+spec:
+  scrapeClass: istio-mtls
+  podMetricsEndpoints:
+  - port: http
+    path: /metrics
+```
+
+The proposed behavior is:
+1. the `tlsConfig` in the associated scrape class is automatically applied to the scrape configuration of the endpoint.
+2. the inline `tlsConfig` (if any) takes precedence over the `tlsConfig` in the scrape class.
```

Comment:
What are we saying?
1. When a tlsConfig field in the pod monitor's endpoint isn't defined, the operator picks up the corresponding value from the scrape class (merge approach).
2. The tlsConfig fields from the scrape class are applied only when the endpoint's tlsConfig is nil.

The latter is simpler but less flexible (e.g. we may want to apply the same CA file to all pod monitors but leave some freedom on the other other TLS params).

Reply:
The first one is what we're saying.
Btw @simonpasquier I've updated the description, wdyt?